src/cipher/mod.rs

```diff
@@ -0,0 +1,359 @@
+///     let mut rng = rand::thread_rng();
+///
+///     // Generate a secret and a public key for Bob
+///     let bob_secret = JubJubScalar::random(&mut rng);
+///     let bob_public = GENERATOR.to_niels().mul(&bob_secret);
+///
+///     // Generate a secret and a public key for Alice
+///     let alice_secret = JubJubScalar::random(&mut rng);
+///     let alice_public = GENERATOR.to_niels().mul(&alice_secret);
+///
+///     // Generate a secret message
+///     let a = BlsScalar::random(&mut rng);
+///     let b = BlsScalar::random(&mut rng);
+///     let message = [a, b];
+///
+///     // Bob's view (sender)
+///     // The cipher and nonce are safe to be broadcasted publicly
+///     let (cipher, nonce) = sender(&bob_secret, &alice_public, &message);
+///
+///     // Alice's view (receiver)
+///     let decrypted_message =
+///         receiver(&alice_secret, &bob_public, &cipher, &nonce);
+///
+///     // Successful communication
+///     assert_eq!(decrypted_message, message);
+/// }
+///
+/// fn sender(
+///     sender_secret: &JubJubScalar,
+///     receiver_public: &ExtendedPoint,
+///     message: &[BlsScalar],
+/// ) -> (PoseidonCipher, BlsScalar) {
+///     // Use the Diffie-Hellman protocol to generate a shared secret
+///     let shared_secret = dhke(sender_secret, receiver_public);
+///
+///     // Generate a random nonce that will be public
+///     let nonce = BlsScalar::random(&mut rand::thread_rng());
+///
+///     // Encrypt the message
+///     let cipher = PoseidonCipher::encrypt(&message, &shared_secret, &nonce);
+///
+///     (cipher, nonce)
+/// }
+///
+/// fn receiver(
+///     receiver_secret: &JubJubScalar,
+///     sender_public: &ExtendedPoint,
+///     cipher: &PoseidonCipher,
+///     nonce: &BlsScalar,
+/// ) -> [BlsScalar; MESSAGE_CAPACITY] {
+///     // Use the Diffie-Hellman protocol to generate a shared secret
+///     let shared_secret = dhke(receiver_secret, sender_public);
+///
+///     // Decrypt the message
+///     cipher
+///         .decrypt(&shared_secret, &nonce)
+///         .expect("Failed to decrypt!")
+/// }
+/// ```
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Default)]
+pub struct PoseidonCipher {
+    cipher: [BlsScalar; CIPHER_SIZE],
+}
+
+impl PoseidonCipher {
+    /// [`PoseidonCipher`] constructor
+    pub fn new(cipher: [BlsScalar; CIPHER_SIZE]) -> Self {
+        Self { cipher }
+    }
+
+    /// Maximum number of scalars allowed per message
+    pub fn capacity() -> usize {
+        MESSAGE_CAPACITY
+    }
+
+    /// Bytes consumed on serialization of the poseidon cipher
+    pub fn serialized_size() -> usize {
+        ENCRYPTED_DATA_SIZE
+    }
+
+    /// Encrypt a slice of scalars into an internal cipher representation
+    ///
+    /// The message size will be truncated to [`MESSAGE_CAPACITY`] bits
+    pub fn encrypt(
+        message: &[BlsScalar],
+        secret: &AffinePoint,
+        nonce: &BlsScalar,
+    ) -> Self {
+        let zero = BlsScalar::zero();
```

Comment:
That has to be constrained as mentioned avobe.

Reply:
This one is not part of the zk circuit